lib/livebook/notebook/explore/kino/pong.livemd

```diff
@@ -1,492 +1,899 @@
+  def on_key(paddle, _), do: paddle
+end
+```
+
+The paddle moves up while the "ArrowUp" key is pressed,
+and down while the "ArrowDown" key is pressed. We will
+render the paddle as a rectangle:
+
+```elixir
+defimpl Scene.Renderer, for: Game.Paddle do
+  def to_svg(%Game.Paddle{x: x, y: y, width: width, height: height}) do
     """
-    <svg viewBox="0 0 #{canvas.width * scale} #{canvas.height * scale}" xmlns="http://www.w3.org/2000/svg">
-      #{Enum.join(object_svgs)}
-    </svg>
+    <rect x="#{x}" y="#{y}" width="#{width}" height="#{height}" fill="black" />
     """
   end
 end
 ```
 
-With this piece of code, we can now declaratively compose an image
-out of basic geometric shapes.
-
-At this point we can already experiment and render our static scene!
+And animate it over the `y` dimension, with the addition
+that the paddle must never leave the scene:
 
 ```elixir
-Canvas.new(400, 200)
-|> Canvas.add_rect(0, 0, 400, 200, fill: "#d4f1f477")
-|> Canvas.add_text(200, 4, "0 : 0", font_size: 10)
-|> Canvas.add_rect(0, 10, 2, 40, fill: "black")
-|> Canvas.add_rect(398, 120, 2, 40, fill: "black")
-|> Canvas.add_circle(200, 100, 8, fill: "tomato")
-|> Canvas.to_svg()
-|> Kino.Image.new(:svg)
+defimpl Animation, for: Game.Paddle do
+  def step(paddle, scene) do
+    %{paddle | y: clip(paddle.y + paddle.dy, 0, scene.height - paddle.height)}
+  end
+
+  defp clip(value, min, max), do: value |> max(min) |> min(max)
+end
 ```
 
-## Game logic
+We are ready to give it a try!
 
-Having the image in mind, we can now model the game state together
-with all its rules. It makes sense to keep this part separate from
-the runtime elements to make it predictable and easy to test.
-Consequently we will encapsulate the whole game state in a struct
-and define functions for the possible interactions.
+<!-- livebook:{"branch_parent_index":4} -->
+
+## Mixing control streams
+
+To animate the paddle, we will need two control streams:
+the keyboard control and an interval to refresh
+the paddle as we keep holding the arrow keys up
+and down. Let's put the relevant pieces in place:
 
 ```elixir
-defmodule Pong do
-  @moduledoc """
-  Represents the paddles game state and rules.
-  """
+# The frame we will render scene images on top of
+frame = Kino.Frame.new() |> Kino.render()
 
-  defstruct [:state, :left, :right, :ball]
+# The scene with a static background
+scene =
+  Scene.new(200, 50)
+  |> Scene.add(Scene.Rect.new(0, 0, 200, 50, fill: "#d4f1f477"))
 
-  @type t :: %__MODULE__{state: state(), left: paddle(), right: paddle(), ball: ball()}
+# The keyboard control
+keyboard = Kino.Control.keyboard([:status, :keyup, :keydown]) |> Kino.render()
 
-  @type state :: :idle | :running | :finished
+# The refresh interval
+interval = Kino.Control.interval(33)
 
-  @type paddle :: %{y: integer(), dy: integer(), points: non_neg_integer()}
+# The paddle which we will animate
+paddle = Game.Paddle.new(0, 21, 2, 8)
```

Comment:
Similarly here, it would be nice to return keyboard/frame, so that the paddle is not inspected.

Reply:
I went with `Kino.render` everywhere for consistency. :)

Reply:
We could also use `Kino.nothing()`.